### ISSUE TITLE
Chart.yaml: kubeVersion not interpreted/compared correctly

### ISSUE THREAD

**Issue by Flauschbaellchen** [helm/helm#10375]
I tried to constrain the Kubernetes version the chart can run on in my `Chart.yaml` with the `kubeVersion` key.
When trying to deploy, the deployment fails due to the comparison not evaluating to true.

Output of `helm version`:
> version.BuildInfo{Version:"v3.7.0", GitCommit:"eeac83883cb4014fe60267ec6373570374ce770b", GitTreeState:"clean", GoVersion:"go1.16.8"}

Output of `kubectl version`:
> Client Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.6-eks-49a6c0", GitCommit:"49a6c0bf091506e7bafcdb1b142351b69363355a", GitTreeState:"clean", BuildDate:"2020-12-23T22:13:28Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}

> Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.13-eks-8df270", GitCommit:"8df2700a72a2598fa3a67c05126fa158fd839620", GitTreeState:"clean", BuildDate:"2021-07-31T01:36:57Z", GoVersion:"go1.15.14", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
> AWS EKS

Chart.yaml (excerpt):
> kubeVersion: ">= 1.19.0"

Error:
> chart requires kubeVersion: >= 1.19.0 which is incompatible with Kubernetes v1.19.13-eks-8df270

In my understanding, the constraint of `>= 1.19.0` should satisfy `v1.19.13-eks-8df270` as `-eks-8df270` is interpreted as metadata and should not be included in the comparison.
Also I checked an online validator, that also showing this constrain to be valid against the k8s version: [[1]]

[1]: https://jubianchi.github.io/semver-check/#/%3E%3D%201.19.0/v1.19.13-eks-8df270

**Comment by bacongobbler** [comment]
A version number with a dash following the `<MAJOR>.<MINOR>.<PATCH>` version fields is marked as a pre-release. Pre-releases are excluded from general semantic version lookups unless the constraint includes a pre-release identifier to the constraint.

There are two ways to fix this:

1. Ask AWS to fix their versioning scheme to match the build metadata schema: `v1.19.6+eks-49a6c0`
2. Change your constraint to include pre-release versions as described in the documentation: `>= 1.19.0-0`

https://helm.sh/docs/chart_template_guide/function_list/#working-with-prerelease-versions

**Comment by Flauschbaellchen** [comment]
Thanks for your response, this was very useful.
I overlooked the difference between + and -.

I opened up an AWS support ticket and will report back as soon as I got an answer if there is the possibility to fix their versioning scheme.

**Comment by bacongobbler** [comment]
In the meantime, I'm going to go ahead and close this issue as answered as I think I provided a workaround in case AWS chooses not to change their versioning scheme. Please feel free to comment here and let us know if you have further questions. Thanks!

**Comment by Flauschbaellchen** [comment]
For anyone who is also interested in the reponse of the AWS team, here is the final answer.
TL;DR: The version scheme of EKS won't be changed in the near feature.

> I would like to inform you that the team has come back with their response and explanation to EKS not following the semvar spec. 

> As I had already assumed, making the version string change is a big change, which is not a priority as yet, and is not likely to be prioritised in the near future either. There have been issues raised previously when EKS accidentally used the build metadata (+) symbol, which was tracked here [1]. In fact, AWS is tracking the importance of this issue here officially, and I would recommend you +1 the ticket to help it gain traction/prioritisation. 

> One thing to highlight, is that the team believes that we need more than 3 version components (major.minor.patch) that semvar has a strict adherence to. As EKS is a redistributor of Kubernetes which already follows the 3 version components from semvar, we require a 4th version component to be able to distinguish which of our multiple versions of EKS using the same Kubernetes version is newer. When using Build metadata (+), you are unable to use a 4th version component to sort which version is newer. However, with Pre-release metadata as you have already seen, there is a workaround in place to include pre-release metadata in the sorting logic. For example:
Using Build Metadata:               1.20.7+eks.12 == 1.20.7+eks.11
Using Pre-Release Metadata:    1.20.7-eks.12 > 1.20.7-eks.11

> Unfortunately, I can't provide an ETA on when EKS team will be revisiting the versioning specifications, but to keep up to date with the latest features announced, take a look at the following websites [2, 3].

> [1] Use correct SemVer for Kubernetes version
> https://github.com/aws/containers-roadmap/issues/1404#issuecomment-884751824 

> [2] Containers Road Map
> https://github.com/aws/containers-roadmap/projects/1 

> [3] EKS News Blog
> https://aws.amazon.com/blogs/aws/category/compute/amazon-kubernetes-service/

**Comment by nfisher** [comment]
Would it make sense to deprecate the current `semverCompare` function and introduce a newly named function? As it is there's something of a cottage industry in how to handle the differences in interpreting AWS's semver format vs the standard. For at least Tier 1 providers (e.g. GCP, Azure, AWS) I would hope Helm would handle their versioning scheme correctly even if it's not semantically correct by semvers standards.

In my experience a lot of the feature toggling is done on the `minor` version as that's typically where Kubernetes will graduate feature flags from off to on. Having natural support baked into Helm such as `>= 1.22` that includes knowledge of Tier 1 cloud providers seems reasonable (AWS at last check is larger than the next 10 IaaS providers combined). Are they bad actors? Maybe, but given the response posted above it feels like there is significant technical obstacles to change to an alternative version label.
